python/ray/data/_internal/execution/operators/streaming_repartition_operator.py

```diff
@@ -0,0 +1,312 @@
+
+        # Initialize the starting index for partitioning the current block
+        start_idx_in_block = 0
+
+        # Continue partitioning the current block until all rows are processed
+        while start_idx_in_block < num_rows_in_block:
+
+            # Calculate the remaining rows to process in the current block
+            remaining_rows_in_block = num_rows_in_block - start_idx_in_block
+
+            # Get the current number of rows in the block being built
+            num_rows_in_cur_block = cur_block_builder.num_rows()
+
+            # If the remaining rows plus the current block's rows do not exceed the
+            # max limit
+            if remaining_rows_in_block + num_rows_in_cur_block < max_num_rows_per_block:
+
+                # Add remaining rows to the current block builder
+                end_idx_in_block = start_idx_in_block + remaining_rows_in_block
+                cur_block_builder.add_block(
+                    accessor.slice(start_idx_in_block, end_idx_in_block, copy=True)
+                )
+                break  # Exit the while loop as no partitioning is needed yet
+
+            # If the current block can exactly match the max row limit
+            elif (
+                remaining_rows_in_block + num_rows_in_cur_block
+                == max_num_rows_per_block
+            ):
+
+                # Partition the current block and create a new one
+                end_idx_in_block = start_idx_in_block + remaining_rows_in_block
+                cur_block_builder.add_block(
+                    accessor.slice(start_idx_in_block, end_idx_in_block, copy=True)
+                )
+
+                # Finalize the partitioned block and store its metadata and content
+                partitioned_block = cur_block_builder.build()
+                metadata = BlockAccessor.for_block(partitioned_block).get_metadata()
+                metadata_list.append(metadata)
+                block_list.append(partitioned_block)
+
+                # Update the starting index for the next partition
+                start_idx_in_block += remaining_rows_in_block
+
+                # Reset the builder for the next partition
+                cur_block_builder = DelegatingBlockBuilder()
+
+            # If the current block exceeds the max row limit, partition it
+            else:
+
+                # Partition the block into smaller pieces based on the max row limit
+                end_idx_in_block = start_idx_in_block + max_num_rows_per_block
+                cur_block_builder.add_block(
+                    accessor.slice(start_idx_in_block, end_idx_in_block, copy=True)
+                )
+
+                # Finalize the partitioned block and store its metadata and content
+                partitioned_block = cur_block_builder.build()
+                metadata = BlockAccessor.for_block(partitioned_block).get_metadata()
+                metadata_list.append(metadata)
+                block_list.append(partitioned_block)
```

Comment:
Let's avoid duplicating the code (for ex, metadata could be collecting at the end, we can lift slicing up, etc)